taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
T.J. @talumbau, I'm totally confused by your comments.

Last week you reported that the redesigned Parameters class had worked fiine with one part of the  webapp, but you couldn't see how to use the redesigned Parameters class in the other part of the webapp (and posted the code you couldn't make work).

Then today, I responded to you suggesting a different way to use the redesigned Parameters class in that second test.  Now in your latest comment in this thread you seem to be saying that the second webapp test fails when using the redesigned Parameters class.  Is that true?  If so, can you post the webapp code that fails when trying to using the redesigned Parameters class and also post the error message you get?  It would be helpful to the OSPC team if you could post a complete report of what testing you have done and what the test results have been.  If there is not a test failure when using the redesigned Parameters class in the webapp, then I don't understand what all the conversation is about.

Can we both argee to focus on the facts of the matter and put off an exchange of opionions until everybody understands the facts?